filters/include/pcl/filters/experimental/voxel_grid.h

```diff
@@ -0,0 +1,513 @@
+      if (filter_field_idx_ == -1) {
+        PCL_WARN("[pcl::%s::applyFilter] Invalid filter field name. Index is %d.\n",
+                 filter_name_.c_str(),
+                 filter_field_idx_);
+      }
+    }
+    else {
+      getMinMax3D<PointT>(*input, *indices, min_p, max_p);
+    }
+
+    min_b_ = (min_p.array() * inverse_leaf_size_).floor().template cast<int>();
+    max_b_ = (max_p.array() * inverse_leaf_size_).floor().template cast<int>();
+
+    div_b_ = (max_b_ - min_b_).array() + 1;
+    divb_mul_ = Eigen::Vector4i(1, div_b_[0], div_b_[0] * div_b_[1], 0);
+
+    num_voxels_ = 0;
+    grid_.clear();
+
+    const boost::optional<std::size_t> max_num_voxels = checkIfOverflow(min_p, max_p);
+    if (max_num_voxels) {
+      grid_.reserve(std::min(max_num_voxels.value(), input->size()));
+    }
+    else {
+      PCL_WARN("[pcl::%s::applyFilter] Leaf size is too small for the input dataset. "
+               "Integer indices would overflow.\n",
+               filter_name_.c_str());
+      return false;
+    }
+
+    if (grid_filter->getSaveLeafLayout()) {
+      const std::size_t new_layout_size = div_b_[0] * div_b_[1] * div_b_[2];
+      const std::size_t reset_size = std::min(new_layout_size, leaf_layout_.size());
+
+      // TODO: stop using std::vector<int> for leaf_layout_ because the range of int is
+      // smaller than size_t
+      std::fill(leaf_layout_.begin(), leaf_layout_.begin() + reset_size, -1);
+
+      try {
+        leaf_layout_.resize(new_layout_size, -1);
+      } catch (std::bad_alloc&) {
+        throw PCLException(
+            "VoxelGrid bin size is too low; impossible to allocate memory for layout",
+            "voxel_grid.hpp",
+            "applyFilter");
+      } catch (std::length_error&) {
+        throw PCLException(
+            "VoxelGrid bin size is too low; impossible to allocate memory for layout",
+            "voxel_grid.hpp",
+            "applyFilter");
+      }
+    }
+
+    return true;
+  }
+
+  boost::optional<std::size_t>
+  checkIfOverflow(const Eigen::Vector4f& min_p, const Eigen::Vector4f& max_p)
+  {
+    // Check that the leaf size is not too small, given the size of the data
+    // Otherwise "wrap around" of unsigned int will happen during hashing a point
+    const std::size_t dx =
+        std::floor((max_p[0] - min_p[0]) * inverse_leaf_size_[0]) + 1;
+    const std::size_t dy =
+        std::floor((max_p[1] - min_p[1]) * inverse_leaf_size_[1]) + 1;
+    const std::size_t dz =
+        std::floor((max_p[2] - min_p[2]) * inverse_leaf_size_[2]) + 1;
+    const size_t max_size = std::numeric_limits<std::size_t>::max();
+
+    // check unsigned int overflow/wrap-around
+    if (dy > max_size / dx || dx * dy > max_size / dz)
+      return boost::none;
+    else
+      return dx * dy * dz;
+  }
+
+  inline std::size_t
+  hashPoint(const PointT& pt)
+  {
+    const std::size_t ijk0 = std::floor(pt.x * inverse_leaf_size_[0]) - min_b_[0];
```

Comment:
It feels like this could be optimized, but I don't have a suggestion yet, will try to think about it.

Reply:
Maybe use `dot` from eigen?

Reply:
I tried hashing with eigen operations, but turns out the original approach is always better. Maybe we should leave it as it is?

[Benchmark with Godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAM1QDsCBlZAQwBtMQBGAFlJvoCqAZ0wAFAB4gA5AAYppAFZdSrZrVDIApACYAQjt2kR7ZATx1KmWugDCqVgFcAtrRDbSV9ABk8tTADlnACNMYhAANncAB1QhQnNaO0cXN1IYuLM6Hz9ApxCwyKNME0zaBgJmYgIk51d3Y0xTBPLKgmyA4NCI%2BoqqmpSe1vbc/O6ASiNUB2JkDikdAGZfZEcsAGpNBZtsPGArAHoAESsRTexNGQBBReXVzA2t5ARiOlQzi%2BvtJdoVh3XNmzmIQEYiYZhOd5XG4/O4PGwAN0aRGIkOuVwccXUa0uxGIzAAntw8A9DmsdntXCAcXj8QDfARSGteGtOO8Fvp0ZjgNjcQTuEISWTdlYQFTebStnEAF6YAD6DKZjNZC3O7I%2BGN83OpfKogvJIrFNO4VE2HLRl18rF892lcoIawQzCECFlmGFtAgaFowJ5RpN2nCayiCq9Pu1hP9gfYzCosttpA%2BayTyZTqbTKdD9vDRJ0gacvllQUZmd9fKEubW%2BHhQVlTgcrDGGwA7GakyXmA4iGsCE4oiTE6moMG1gAqNbR2O2sYAOiorFQJAgM5YwLp9DOS7WAFo1vnaIXl06CADbfKN2NTQPQQRprQ1hAe32x1Wa3WG9OEGD0ACFufp0JnCXS8oSbQ4Pg%2BS1rTWU97UdZ1ZRIXZfDYT06DDcVjQrYNizQrMMMjccwUnPAZQTK50woyiM1w0tCTwCs90LHDvTwo1ywDSs8GrWt60bTQWwHEsYLWPAFAAaxkQVgXQUU5wXYgIGDTQAFZdBkFTSTHCc4xIzAVLUjTGx3Rign09TlLAtVyLWITdPlETxM4KSCBkkA5MXJTVNZCzR0ImMdJlfTvMOIzdwLUyvI04DLjbGjhNEsTtGc1z3IUzzdG0DTfO0219MyizQpMvKoqsmK1mvW8HIk3yXx41gzKygwqqc58uNfesgsavQqqS1ruLfYqLOi/iwKhK56V3ZhfE3fjWxsuK7PtWhnFlQhQmYUoBU2UlOBkPb9vU0qB2zXUct0wUTtFAAlNR0FQJwgNKpNs2JEyIE4Rl3DWBZGW4C8nto/lOP6%2BsIGUxlwkZJtGQADn%2Bs0B2k0VEVMEgAROs4gwIcsAeDIRp1BERiERCBlqcVaCHWzb4YHGhiHveLBUO3QRLhNYyYpqmEhxlmDAMPA%2BIE6ykzx6cogxF0gmYZAxIgS6QBu6x7qXGnrJG8DrKdImCAgOCXTdClFOxhqLMZM7SLC/ci2B9qG37BZRrK9M9YQ4gkNoFC8ZNw4zaIgLMEZEzGVqt8xlV80kw7LsCCc7boJc0UnheWhUFFYEv3xWUVlQGXRRTgB3R65rphnFtZuPmfLrZ2ZWta8U200Nj0fnBbmpMXYNqwjfLVT6NNvziItoObbq8Ok3VzXO1QbskrjpGUGeV408p5h0Ez7Pc5AAui41sqo%2BnggFmSxPF5T5eM6z%2BdN%2B3se1hLiBGYrxviQBGvybrjbucbvm9AF5s24dE6F0iFgDIVYN3fSfcfYD39oHcKwc2qj2iuPUCA597dm4MfBeydU4gHTqvdeV8xJ51QIXcOiME4oCmPaAEr8dDaE7q4JumVhaUVodXYEn9kBZyPCeShScl4gHQNMT%2BdAAR3QcEEdgbJsAPiSjuGOy4pj0CAqwii7CbDx1cp4eqAN55oE7HCOh2htAgLASAZhA4qIaPjlwnhq5JT8NPrg4R9cEjiKmFIvSKoNwEEwQohYSiHAqIvGo9MNj546OQQ8BG1kKrEDvMzcCoEpATFYNIZS8hXCyHkKnKQNgf4syEFMGY9xFicHkAQaQcgw6kGIbtGQKhpDcHkE4CITZpycAAJzaE4DDGQykZBdMiDDbgXTSDZLkKQPJ8ghAgEaVUnJEw4CwCQJgcQjQp5kAoBASowAhCiDUMUBgCBSHZIqaQNAvY8CqFKIcvwrATlnOqfIK5UQbldGAJwb5lz7rvPYMQAA8p2J5%2BdJnyHWY0HE%2BzpAQo2cgco%2BBsnyH4IwFg7AuC8FRcIMQkgcmKGUKodQKBCkqDwEEOZsAPZOA4Iw0giIwhSAWAsLpqSJioGDNzaQW4tzSW2loZuegeDbkBQsWZJTZjKGkr4e5xzTlgukBc/OeIoiKtSU0qQmSJkvOmdIcQMNwhbnCJg4AyBkAsk4J0sk8Ktn3lwIQEgTcFicEbAUwVuhKkvNqekqQLTSBtIWNoac/Smww26WMmQCwmycGUu4cFuqpCzPmaQRZNS0nSDFdq/FMyU1eomAyuIFhuBAA%3D)

Reply:
Some more thoughts on this (correct me if I'm wrong somewhere):
- Since `divb_mul[0]` is always 1, multiplication with it seems wasted
- Are the `Eigen::Vector4i` sufficient or do we need vectors/arrays of uint32_t or uint64_t to not run into limitations at some point in the future?
- To me it would seem more natural if we first subtract the minimum corner of the bounding box and then multiply by the inverse leaf size, instead of first multiplying then subtracting `min_b`. That might also prevent some overflows, but not sure if there are downsides?
- The Godbolt benchmark is not *that* meaningful, I suspect that some code parts are optimized away by the compiler. The google benchmark tool has some magic to prevent that. If you like, see how fast the google benchmark you already wrote is, when you swap out the hashing function

Reply:
- Agree, I'll remove it
- With the current implementation it is sufficient (not sufficient if we want to use eigen operations above)
- No downside except for storing 2 more vectors as member
- I tested with the pcl filter benchmark, the run time are nearly the same on my machine. I'll keep it for current implementation for now

Reply:
If runtime is same, and eigen is more terse and easy to read, I vote for that. But for now, it's ok